File: server/src/algo2_py/algo.py
from container import get_score, update_pps
from box_motion import rotation, perturbation
from metrics import order_metric, overall_metric
from box import Box
import copy
import sys
import json


def get_best_point(pp, box, container, solutionBoxes):
    bestPoint = None
    bestScore = [0, 0]
    
    for p in pp:
        score = get_score(box, p, solutionBoxes, container)
        if score[0] > bestScore[0] or (score[0] == bestScore[0] and score[1] > bestScore[1]):
            bestPoint = p
            bestScore = score
    
    return bestPoint


def add_box_to_solution(bestPoint, box, pp, solutionBoxes, solution_data, retryList, container, isRetry):
    if bestPoint:
        box.set_position(bestPoint)
        pp = update_pps(box, bestPoint, pp, container)
        solutionBoxes.append(box)
        solution_data["number_of_items"] += 1
        solution_data["capacity"] += box.volume
    else:
        if not isRetry:
            retryList.append(box)
        else:
            solutionBoxes.append(box)

def handle_box(box, pp, container, solutionBoxes, solution_data, retryList, isRetry):
    bestPoint = get_best_point(pp, box, container, solutionBoxes)
    add_box_to_solution(bestPoint, box, pp, solutionBoxes, solution_data, retryList, container, isRetry)


def constructive_packing(boxes, container, isQuantity):
    pp = {
        (0, 0, 0, 1),
        (container["width"], 0, 0, -1)
    }
    
    solutionBoxes = []
    retryList = []
    
    solution_data = {
        "number_of_items": 0,
        "capacity": 0,
        "order_score": 0,
        "overall_score": 0
    }
    
    for box in boxes:
        handle_box(box, pp, container, solutionBoxes, solution_data, retryList, False)
    
    for box in retryList:
        handle_box(box, pp, container, solutionBoxes, solution_data, retryList, True)
    
    solution_data["order_score"] = float(order_metric(solutionBoxes, container))
    solution_data["overall_score"] = float(overall_metric(boxes, container, solution_data, isQuantity))
    
    return [solutionBoxes, solution_data]

def handle_data(data):
    isQuality = int(data['project_data']['isQuality'])
    numOfIterations = 2000 if isQuality else 1000
    isQuantity = int(data['project_data']['isQuantity'])
    container = data['container']
    boxes = data['boxes']

    initBoxes = [Box(box) for box in boxes]
    initBoxes.sort(key=lambda b: b.order, reverse=True)

    return initBoxes, container, numOfIterations, isQuantity

def get_solutions(numOfIterations, boxes, container, isQuantity):
    solutionList = {}
    counter = 0
    for _ in range(numOfIterations):
        #TODO: copy or deepcopy?
        boxesCopy = copy.deepcopy(boxes)
        rotation(boxesCopy)
        perturbation(boxesCopy)
        solution = constructive_packing(boxesCopy, container, isQuantity)
        if solution is None:
            continue
        boxesInSolution, solution_data = solution
        if boxesInSolution is not None and solution_data is not None:
            ## currentBoxes = [box.get_box() for box in boxesInSolution]
            counterString = str(counter)
            solutionList[counterString] = {
                "name": "solution " + counterString,
                "id": counter,
                "boxes": boxesInSolution, ##currentBoxes,
                "solution_data": solution_data
            }
            counter += 1
    return solutionList


def sort_by_preference(solutionList, isQuantity):
    if isQuantity:
        solutionList = sorted(solutionList.values(),
                              key=lambda x: (x["solution_data"]["number_of_items"], x["solution_data"]["capacity"], -x["solution_data"]["order_score"]),
                              reverse=True)[:50]
        solutionList = sorted(solutionList,
                              key=lambda x: (x["solution_data"]["capacity"]),
                              reverse=True)[:25]
        solutionList = sorted(solutionList,
                              key=lambda x: (x["solution_data"]["order_score"]),
                              reverse=False)[:10]
    else:
        solutionList = sorted(solutionList.values(),
                              key=lambda x: (x["solution_data"]["order_score"]),
                              reverse=False)[:20]
        solutionList = sorted(solutionList,
                              key=lambda x: (x["solution_data"]["number_of_items"], x["solution_data"]["capacity"]),
                              reverse=True)[:15]
        solutionList = sorted(solutionList,
                              key=lambda x: (x["solution_data"]["capacity"]),
                              reverse=True)[:10]
    solutionList = sorted(solutionList,
                          key=lambda x: (x["solution_data"]["overall_score"]),
                          reverse=True)
    return solutionList


def dict_solutions_from_list(solutionList):
    solutionDict = {}
    for index, solution in enumerate(solutionList):
        indexString = str(index)
        solutionDict[indexString] = solution
    return solutionDict


def algo():
    data = json.loads(sys.argv[1])

    boxes, container, numOfIterations, isQuantity = handle_data(data)
  
    solutionList = get_solutions(numOfIterations, boxes, container, isQuantity)
    solutionList = sort_by_preference(solutionList, isQuantity)
    solutionDict = dict_solutions_from_list(solutionList)

    return solutionDict

print(algo())

File: server/src/algo_js/algo.js
const { initBox, setPosition, getBox } = require("./box.js");
const { getScore, updatePps } = require("./container.js");
const { rotation, perturbation } = require("./boxMotion.js");
const { orderMetric, overallMetric } = require("./metrics.js");

const ALGOIRTHM_MAX_ITERATIONS = 1000000;

const getBestPoint = (pp, box, container, solutionBoxes) => {
	let bestPoint = null;
	let bestScore = [0, 0];

	pp.forEach((p) => {
		let score = getScore(box, p, solutionBoxes, container);
		if (
			score[0] > bestScore[0] ||
			(score[0] === bestScore[0] && score[1] > bestScore[1])
		) {
			bestPoint = p;
			bestScore = score;
		}
	});

	return bestPoint;
};

const addBoxToSolution = (
	bestPoint,
	box,
	pp,
	solutionBoxes,
	solution_data,
	retryList,
	container,
	isRetry
) => {
	if (bestPoint) {
		setPosition(box, bestPoint);
		pp = updatePps(box, bestPoint, pp, container);
		solutionBoxes.push(box);
		solution_data["number_of_items"] += 1;
		solution_data["capacity"] += box.volume;
	} else {
		if (!isRetry) {
			retryList.push(box);
		} else {
			solutionBoxes.push(box);
		}
	}
};

const handleBox = (
	box,
	pp,
	container,
	solutionBoxes,
	solution_data,
	retryList,
	isRetry
) => {
	const bestPoint = getBestPoint(pp, box, container, solutionBoxes);
	addBoxToSolution(
		bestPoint,
		box,
		pp,
		solutionBoxes,
		solution_data,
		retryList,
		container,
		isRetry
	);
};

const constructivePacking = (boxes, container, isQuantity) => {
	/*
	Initialize potential points as the FLB and FRB of the container.
	Each point has a direction that indicates which corner of a box should be at the point.
	1 = FLB, -1 = FRB.
	*/
	let pp = new Set([
		{
			x: 0,
			y: 0,
			z: 0,
			dir: 1,
		},
		{
			x: container.width,
			y: 0,
			z: 0,
			dir: -1,
		},
	]);

	let solutionBoxes = [];
	let retryList = [];

	let solution_data = {
		number_of_items: 0,
		capacity: 0,
		order_score: 0,
		overall_score: 0,
	};

	// For each box, put it in the best point possible.
	boxes.forEach((box) => {
		handleBox(
			box,
			pp,
			container,
			solutionBoxes,
			solution_data,
			retryList,
			false
		);
	});

	// For each box which is not put in the solution, try and put it again.
	retryList.forEach((box) => {
		handleBox(
			box,
			pp,
			container,
			solutionBoxes,
			solution_data,
			retryList,
			true
		);
	});

	solution_data.order_score = parseFloat(
		orderMetric(solutionBoxes, container)
	);
	solution_data.overall_score = parseFloat(
		overallMetric(boxes, container, solution_data, isQuantity)
	);
	return [solutionBoxes, solution_data];
};

const handleData = (data) => {
	const isQuality = data.project_data.isQuality;
	const algorithmTime = isQuality ? 60000 : 15000; // miliseconds
	const isQuantity = data.project_data.isQuantity;
	const container = data.container;
	const boxes = data.boxes;

	const initBoxes = boxes.map((box) => {
		return initBox(box);
	});

	// Sort the boxes by descending order by the "order" property.
	initBoxes.sort((a, b) => (a.order > b.order ? -1 : 1));
	return [initBoxes, container, isQuantity, algorithmTime];
};

const getSolutions = (algorithmTime, boxes, container, isQuantity) => {
	let solutionList = {};
	let counter = 0;
	const endTime = Date.now() + algorithmTime;
	while (Date.now() < endTime) {
		let boxesCopy = [...boxes];
		// For non-deterministic algorithm - rotating and perturbating the boxes.
		boxesCopy = rotation(boxesCopy);
		boxesCopy = perturbation(boxesCopy);
		let solution = constructivePacking(boxesCopy, container, isQuantity);
		if (solution === null) {
			continue;
		}
		let [boxesInSolution, solution_data] = solution;
		if (boxesInSolution !== null && solution_data !== null) {
			let counterString = counter.toString();
			solutionList[counterString] = {
				name: "solution " + counterString,
				id: counter,
				boxes: boxesInSolution.map((box) => {
					return getBox(box);
				}),
				solution_data: solution_data,
			};
			counter += 1;
		}
		if (counter === ALGOIRTHM_MAX_ITERATIONS) {
			break;
		}
	}

	return solutionList;
};

const sortByPreference = (solutionList, isQuantity) => {
	if (isQuantity) {
		solutionList = Object.values(solutionList)
			.sort(
				(a, b) =>
					b.solution_data.number_of_items -
					a.solution_data.number_of_items
			)
			.slice(0, 50);
		solutionList = solutionList
			.sort((a, b) => b.solution_data.capacity - a.solution_data.capacity)
			.slice(0, 25);
		solutionList = solutionList
			.sort(
				(a, b) =>
					b.solution_data.order_score - a.solution_data.order_score
			)
			.slice(0, 10);
	} else {
		solutionList = Object.values(solutionList)
			.sort(
				(a, b) =>
					b.solution_data.order_score - a.solution_data.order_score
			)
			.slice(0, 20);
		solutionList = solutionList
			.sort(
				(a, b) =>
					b.solution_data.number_of_items -
					a.solution_data.number_of_items
			)
			.slice(0, 15);
		solutionList = solutionList
			.sort((a, b) => b.solution_data.capacity - a.solution_data.capacity)
			.slice(0, 10);
	}
	solutionList = solutionList.sort(
		(a, b) => b.solution_data.overall_score - a.solution_data.overall_score
	);
	return solutionList;
};

const dictSolutionsFromList = (solutionList) => {
	let solutionDict = {};
	for (let index = 0; index < solutionList.length; index++) {
		let indexString = index.toString();
		solutionDict[indexString] = solutionList[index];
	}
	return solutionDict;
};

const algo = (data) => {
	const [boxes, container, isQuantity, algorithmTime] = handleData(data);
	let solutionList = getSolutions(
		algorithmTime,
		boxes,
		container,
		isQuantity
	);

	solutionList = sortByPreference(solutionList, isQuantity);

	solutionList = solutionList.map((solution, index) => {
		return { ...solution, name: `solution ${index + 1}` };
	});

	let solutionDict = dictSolutionsFromList(solutionList);

	return solutionDict;
};

module.exports = { algo, handleBox };
